src/main/java/org/reaktivity/nukleus/http_cache/internal/proxy/cache/DefaultCache.java

```diff
@@ -144,24 +169,96 @@ public boolean matchCacheableRequest(
         short authScope,
         int requestHash)
     {
-        final DefaultCacheEntry cacheEntry = cachedEntries.get(requestHash);
+        final DefaultCacheEntry cacheEntry = cachedEntriesByRequestHash.get(requestHash);
         return satisfiedByCache(requestHeaders) &&
                 cacheEntry != null &&
-                cacheEntry.canServeRequest(requestHeaders, authScope);
-
+                cacheEntry.canServeRequest(requestHeaders, authScope) &&
+                cacheEntry.isValid();
     }
 
     public void purge(
         int requestHash)
     {
-        DefaultCacheEntry cacheEntry = this.cachedEntries.remove(requestHash);
-        entryCount.accept(-1);
+        DefaultCacheEntry cacheEntry = cachedEntriesByRequestHash.remove(requestHash);
         if (cacheEntry != null)
         {
+            final int requestHashWithoutQuery = cacheEntry.requestHashWithoutQuery();
+            Int2ObjectHashMap cachedEntriesByRequestHashFromWithoutQueryList = cachedEntriesByRequestHashWithoutQuery
+                .computeIfPresent(requestHashWithoutQuery, (h, m) -> m.remove(requestHash) != null && m.isEmpty() ? null : m);
+            if (cachedEntriesByRequestHashFromWithoutQueryList == null)
+            {
+                cachedEntriesByRequestHashWithoutQuery.remove(requestHashWithoutQuery);
+            }
+
+            entryCount.accept(-1);
             cacheEntry.purge();
         }
     }
 
+    public void invalidateCacheEntryIfNecessary(
+        HttpCacheProxyFactory factory,
+        int requestHash,
+        String requestURL,
+        ArrayFW<HttpHeaderFW> headers)
+    {
+        DefaultCacheEntry cacheEntry = cachedEntriesByRequestHash.remove(requestHash);
+        if (cacheEntry != null)
+        {
+            cacheEntry.invalidate();
+        }
+
+        headers.forEach(header ->
+        {
+            invalidateLinkCacheEntry(factory, requestURL, header);
+        });
```

Comment:
Please make this an inline lambda instead, removing curly braces.

Reply:
Done.